File: lib/features/account/data/models/warning.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'ban.dart'; // For RelatedContent class

enum WarningType {
  content_violation,
  inappropriate_behavior,
  spam,
  harassment,
  other,
}

enum WarningSeverity {
  low,
  medium,
  high,
  critical,
}

class Warning {
  final String id;
  final String userId;
  final WarningType type;
  final String reason;
  final String? description;
  final WarningSeverity severity;
  final String issuedBy; // Admin UID or email
  final DateTime issuedAt;
  final bool isActive;
  final List<String>? deviceIds; // User's device IDs at time of warning
  final RelatedContent? relatedContent;
  final String? reportId; // Link to user report if applicable

  const Warning({
    required this.id,
    required this.userId,
    required this.type,
    required this.reason,
    this.description,
    required this.severity,
    required this.issuedBy,
    required this.issuedAt,
    required this.isActive,
    this.deviceIds,
    this.relatedContent,
    this.reportId,
  });

  factory Warning.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return Warning(
      id: doc.id,
      userId: data['userId'] as String,
      type: WarningType.values.firstWhere(
        (e) => e.name == data['type'],
        orElse: () => WarningType.other,
      ),
      reason: data['reason'] as String,
      description: data['description'] as String?,
      severity: WarningSeverity.values.firstWhere(
        (e) => e.name == data['severity'],
        orElse: () => WarningSeverity.low,
      ),
      issuedBy: data['issuedBy'] as String,
      issuedAt: (data['issuedAt'] as Timestamp).toDate(),
      isActive: data['isActive'] as bool? ?? true,
      deviceIds: data['deviceIds'] != null
          ? List<String>.from(data['deviceIds'])
          : null,
      relatedContent: data['relatedContent'] != null
          ? RelatedContent.fromMap(data['relatedContent'])
          : null,
      reportId: data['reportId'] as String?,
    );
  }

  Map<String, dynamic> toFirestore() {
    return {
      'userId': userId,
      'type': type.name,
      'reason': reason,
      'description': description,
      'severity': severity.name,
      'issuedBy': issuedBy,
      'issuedAt': Timestamp.fromDate(issuedAt),
      'isActive': isActive,
      'deviceIds': deviceIds,
      'relatedContent': relatedContent?.toMap(),
      'reportId': reportId,
    };
  }

  Warning copyWith({
    String? id,
    String? userId,
    WarningType? type,
    String? reason,
    String? description,
    WarningSeverity? severity,
    String? issuedBy,
    DateTime? issuedAt,
    bool? isActive,
    List<String>? deviceIds,
    RelatedContent? relatedContent,
    String? reportId,
  }) {
    return Warning(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      type: type ?? this.type,
      reason: reason ?? this.reason,
      description: description ?? this.description,
      severity: severity ?? this.severity,
      issuedBy: issuedBy ?? this.issuedBy,
      issuedAt: issuedAt ?? this.issuedAt,
      isActive: isActive ?? this.isActive,
      deviceIds: deviceIds ?? this.deviceIds,
      relatedContent: relatedContent ?? this.relatedContent,
      reportId: reportId ?? this.reportId,
    );
  }
}

File: apps/ios/ios/Core/Models/Warning.swift
import Foundation
import FirebaseFirestore

/// Ported from: apps/mobile/lib/features/account/data/models/warning.dart

enum WarningType: String, Codable {
    case content_violation
    case inappropriate_behavior
    case spam
    case harassment
    case other
}

enum WarningSeverity: String, Codable {
    case low
    case medium
    case high
    case critical
}

struct Warning: Identifiable {
    let id: String
    let userId: String
    let type: WarningType
    let reason: String
    let description: String?
    let severity: WarningSeverity
    let issuedBy: String
    let issuedAt: Date
    let isActive: Bool
    let deviceIds: [String]?
    let relatedContent: RelatedContent?
    let reportId: String?

    static func fromFirestore(_ document: DocumentSnapshot) -> Warning? {
        guard let data = document.data() else { return nil }

        return Warning(
            id: document.documentID,
            userId: data["userId"] as? String ?? "",
            type: WarningType(rawValue: data["type"] as? String ?? "") ?? .other,
            reason: data["reason"] as? String ?? "",
            description: data["description"] as? String,
            severity: WarningSeverity(rawValue: data["severity"] as? String ?? "") ?? .low,
            issuedBy: data["issuedBy"] as? String ?? "",
            issuedAt: (data["issuedAt"] as? Timestamp)?.dateValue() ?? Date(),
            isActive: data["isActive"] as? Bool ?? true,
            deviceIds: data["deviceIds"] as? [String],
            relatedContent: {
                guard let rc = data["relatedContent"] as? [String: Any] else { return nil }
                return RelatedContent(
                    type: rc["type"] as? String ?? "",
                    id: rc["id"] as? String ?? "",
                    title: rc["title"] as? String,
                    metadata: rc["metadata"] as? [String: String]
                )
            }(),
            reportId: data["reportId"] as? String
        )
    }
}
